src/platform/w3dengine/client/w3dtreebuffer.cpp

```diff
@@ -0,0 +1,1553 @@
+                            float sin = m_trees[j].sin;
+                            float cos = m_trees[j].cos;
+
+                            if (tree_type < 0 || m_treeTypes[tree_type].mesh == nullptr) {
+                                tree_type = 0;
+                            }
+
+                            VertexFormatXYZNDUV1 *current_vertex = &vertexes[m_trees[j].vertex_array_index];
+                            int vert_count = m_treeTypes[tree_type].mesh->Peek_Model()->Get_Vertex_Count();
+                            Vector3 *vert_array = m_treeTypes[tree_type].mesh->Peek_Model()->Get_Vertex_Array();
+
+                            for (int k = 0; k < vert_count; k++) {
+                                float vx = vert_array[k].X;
+                                float vy = vert_array[k].Y;
+                                Vector3 vertex;
+                                vx = vx + m_treeTypes[tree_type].pos.X;
+                                vy = vy + m_treeTypes[tree_type].pos.Y;
+                                vertex.X = vx * scale * cos - vy * scale * sin;
+                                vertex.Y = vy * scale * cos + vx * scale * sin;
+                                vertex.Z = scale * vert_array[k].Z;
+                                vertex.Z = vertex.Z + m_treeTypes[tree_type].pos.Z;
+
+                                if (m_trees[j].topple_state) {
+                                    Matrix3D::Transform_Vector(m_trees[j].topple_transform, vertex, &vertex);
+                                } else {
+                                    if (m_trees[j].push_aside_speed > 0.0f) {
+                                        vertex.X = vert_array[k].Z * m_trees[j].push_aside_speed
+                                                * m_trees[j].push_aside_location.y
+                                                * m_treeTypes[tree_type].module->m_moveOutwardDistanceFactor
+                                            + vertex.X;
+                                        vertex.Y = vert_array[k].Z * m_trees[j].push_aside_speed
+                                                * m_trees[j].push_aside_location.x
+                                                * m_treeTypes[tree_type].module->m_moveOutwardDistanceFactor
+                                            + vertex.Y;
+                                    }
+
+                                    vertex.X = vertex.X + location.X;
+                                    vertex.Y = vertex.Y + location.Y;
+                                    vertex.Z = vertex.Z + location.Z;
+                                }
+
+                                current_vertex->x = vertex.X;
+                                current_vertex->y = vertex.Y;
+                                current_vertex->z = vertex.Z;
+                                current_vertex->ny =
+                                    1.0f - m_treeTypes[tree_type].module->m_darkeningFactor * m_trees[j].push_aside_speed;
```

Comment:
comment as possible bug, use of m_darkeningFactor is odd here

Reply:
fixed.